### DISCUSSION TITLE
Discussion: Well-known member `Struct.IsDefault`

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#6660] [Language Ideas]
Even though list-patterns can operate on types like `ImmurableArray<T>`, the usage is not safe in case of a default value.
```cs
ImmutableArray<T> array = default;
if (array is []) // throws NRE
```
A possible solution is to simply check `IsDefault` as part of the match. Additionally, it could work with `??` operators as well.
```cs
return array ?? [1, 2, 3];
```
This would be a breaking change unless there's a requirement to annotate such properties or a different syntax is used.

Note: this is NOT a proposal to extend nullability analysis (see https://github.com/dotnet/csharplang/discussions/5337).

**Comment by svick** [comment]
My problem with this approach is that it's effectively a special case just for a single type: `ImmutableArray<T>`. While in theory it would apply to any type that follows the same pattern, I don't think those are common.

**Reply by alrz** [reply]
Among `struct` types in BCL `IsDefault` is used only for this purpose. https://source.dot.net/#q=IsDefault

I'd like to think that's a universal convention but if that's not the case, an attribute could certainly be used.

**Comment by alrz** [comment]
Even though this will be known as a new "null check" I don't think this should allow `array is not null` on structs. Rather, it could be emitted using `array is not default` (which is invalid today) in addition to recursive and list-patterns. Question is, what should that do for types without `IsDefault` - error?

**Comment by RikkiGibson** [comment]
I don't know exactly what to do here but I do strongly feel that the quality of life issues for ImmutableArray ought to be addressed. I do agree that if the compiler introduces some recognition of "is it default" which is picked up by patterns, then it should be threaded into most of the language constructs which do null tests: `??`, `?.`, etc.

What happens if we do something like:

```cs
ImmutableArray<int> arr = default;
M(arr);

void M<T>(T t) where T : IReadOnlyList<int>
{
    if (t is []) { }
}
```

We might have to continue to throw at runtime here. It would be a big improvement to do the right thing when the type is known statically, just too bad to have these corner cases remain jagged.

**Reply by qrli** [reply]
My humble opinion is that: uninitialized `ImmutableArray<T>` should not throw NRE at all, instead it should behave as empty collection. It is more issue of the design choice of Immutable collection types. Its `IReadOnlyList<T>` implementation should just work. 

This is not pattern-matching only issue. I don't think there are code taking a non-null `IReadOnlyList<T>` or `IEnumerable<T>` defending against NRE at all.

**Reply by CyrusNajmabadi** [reply]
> My humble opinion is that: uninitialized ImmutableArray<T> should not throw NRE 

Then it would not be a drop in for arrays. And there would be no way to indicate a non-initialized array versus an empty array.

**Reply by qrli** [reply]
@CyrusNajmabadi Could you elaborate? What do you mean by `non-initialized array`?
My understanding is that, an array variable is either null or an object which can be used as `IReadOnlyList<T>` without NRE.

I know there are unsafe runtime APIs to create array without initializing the elements. If you meant that, I'd argue that's not what most .NET developers use, or even know about. And `unsafe` already tells that the developer using it shall take full responsibility. 

If you mean the equivalence of `null`, well, ... C# already has nullable value type, why invent yet another null state.

**Reply by RikkiGibson** [reply]
IIRC, early implementations of ImmutableArray actually treated the default value like an empty array. But it was felt that the extra null checks required to do this were not acceptable.

**Reply by qrli** [reply]
I'm confused. By 'empty array', do you mean `[]` or `null`?
If the former, there should be no null check.
If the latter, having `IsDefault` and throwing NRE is just another type of `null`, and worse.

**Reply by CyrusNajmabadi** [reply]
> Could you elaborate? What do you mean by non-initialized array?

An uninitialized immutable array. This needs to be the same as an uninitialized array (same semantics, same perf), out it's not a drop in replacement. 

An uninitialized array is not empty. So the same should be true for an immutable array.

**Reply by qrli** [reply]
@CyrusNajmabadi OK. I get what you are referring to.
Though, I still think `IsDefault` behaves more like null than uninitialized array.
```cs
// null
string[]? a = default;
Console.WriteLine(a.Count()); // ArgumentNullException
// uninitialized
string[] b = new string[10]; 
Console.WriteLine(b.Count()); // 10

// uninitialized (but behaves more like null array IMO)
ImmutableArray<string> c = default;
Console.WriteLine(c.Count()); // InvalidOperationException

// not so pretty with nullable
ImmutableArray<string>? d = null;
Console.WriteLine(d.Count()); // Not compile; must write d.Value.Count()
```

**Reply by CyrusNajmabadi** [reply]
> `ImmutableArray<string>?`

This would not be a drop in replacement.  It would increase the size needed to move from an array to an ImmutableArray.  

> // uninitialized (but behaves more like null array IMO)

The point is to have tehse two fields behave as close to each other as possible:

```c#
class C
{
    X[] anArray;
    ImmutableArray<X> anImmutableArray;
}
```

The goal was to ensure that code that previously was using the first form could move to the second form without things like extra memory being used.  That means that just like you can check if `anArray` is `null`, and have that be distinct from an empty array, you also need the same for ImmutableArray (without the overhead of something like nullable).  That's also why IA overloads `==` and actually will return `true` if you compare the `default` value of it to `null`.

It's an array replacement.

**Reply by qrli** [reply]
@CyrusNajmabadi Yes, I get that.
And I agree that `ImmutableArray<string>?` is not a good enough alternative for the null state in this case.

Yet still, IA still created yet another null-like state, which does not behave exactly like a null array, in my examples above and the examples by others in this thread. And there could be much more examples: e.g., with NRT `X[]` give some level non-null check (or at least express the intention of non-null), while the compiler does not help the custom null-like state.

So, I think this point still holds: this is not a pattern matching only issue - it is essentially about the language to support custom null-like state.
